titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java

```diff
@@ -0,0 +1,129 @@
+import com.netflix.titus.common.network.client.RxHttpResponse;
+import com.netflix.titus.common.network.client.RxRestClient;
+import com.netflix.titus.common.network.client.RxRestClientException;
+import com.netflix.titus.common.network.client.RxRestClients;
+import com.netflix.titus.common.network.client.TypeProviders;
+import com.netflix.titus.common.util.guice.ProxyType;
+import com.netflix.titus.common.util.guice.annotation.ProxyConfiguration;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.ssl.SslContextBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import rx.Observable;
+import rx.Single;
+
+/**
+ * This {@link DefaultDockerRegistryClient} implementation of {@link RegistryClient} connects to a
+ * Docker V2 REST API compatible registry endpoint.
+ */
+@Singleton
+@ProxyConfiguration(types = {ProxyType.Logging, ProxyType.Spectator})
+public class DefaultDockerRegistryClient implements RegistryClient {
+
+    private static final Logger logger = LoggerFactory.getLogger(DefaultDockerRegistryClient.class);
+
+    private static final String acceptHeader = "Accept";
+    private static final String dockerManifestType = "application/vnd.docker.distribution.manifest.v2+json";
+    private static final String dockerDigestHeaderKey = "Docker-Content-Digest";
+
+    private static final Map<String, String> headers = Collections.unmodifiableMap(
+            Stream.of(new AbstractMap.SimpleEntry<>(acceptHeader, dockerManifestType))
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue))
+    );
+
+    private final TitusRegistryClientConfiguration titusRegistryClientConfiguration;
+    private final RxRestClient restClient;
+
+    @Inject
+    DefaultDockerRegistryClient(RegistryEndpointResolver endpointResolver, TitusRegistryClientConfiguration configuration, Registry spectatorRegistry) {
+        this.titusRegistryClientConfiguration = configuration;
+
+        RxRestClients.Builder builder = RxRestClients.newBuilder("dockerRegistryClient", spectatorRegistry)
+                .endpointResolver(endpointResolver)
+                // We rely on the RxRestClient for timeout handling
+                .timeUnit(TimeUnit.MILLISECONDS)
+                .requestTimeout(titusRegistryClientConfiguration.getRegistryTimeoutMs())
+                .retryCount(titusRegistryClientConfiguration.getRegistryRetryCount())
+                .retryDelay(titusRegistryClientConfiguration.getRegistryRetryDelayMs())
+                .noRetryStatuses(new HashSet<>(Collections.singletonList(HttpResponseStatus.NOT_FOUND)));
+        if (titusRegistryClientConfiguration.isSecure()) {
+            try {
+                builder.sslEngineFactory(new NettySslContextEngineFactory(SslContextBuilder.forClient().build()));
+            } catch (SSLException e) {
+                logger.error("Unable configure Docker registry client SSL context: {}", e.getMessage());
```

Comment:
add `, e` at the end to get full stack trace

Reply:
Will do.